src/Grpc/JsonTranscoding/src/Shared/ServiceDescriptorHelpers.cs

```diff
@@ -96,6 +100,34 @@ public static bool TryResolveDescriptors(MessageDescriptor messageDescriptor, IL
         return fieldDescriptors != null;
     }
 
+    private static FieldDescriptor? GetFieldByName(MessageDescriptor messageDescriptor, string fieldName)
+    {
+        // Search fields by field name and JSON name.
+        // JSON name takes precedence. If there are conflicts then the last field with a name wins.
+        // This logic matches how properties are used in JSON serialization's MessageTypeInfoResolver.
+        var fields = messageDescriptor.Fields.InFieldNumberOrder();
+
+        FieldDescriptor? fieldDescriptor = null;
+        for (var i = fields.Count - 1; i >= 0; i--)
+        {
+            // We're checking JSON name first, in reverse order through fields.
+            // That means the method can exit early on match because the match has the highest precedence.
```

Comment:
comment verified against `CreateJsonFieldMap`, however: I'm not sure that `CreateJsonFieldMap` has the correct implementation; thought experiment:

- field `a`, json name 'b'
- field `b`, json name 'c`
- field `c`, json name `a`

(yes, this is deliberately evil)

with the current logic, this will have the result that field c claims both `c` and `a`, and field b takes `b`, when IMO that is the wrong outcome if JSON names are meant to take precedence

```
// scenario
var fieldsInOrder = new (string Name, string JsonName)[] { ("a", "b"), ("b", "c"), ("c", "a") };

// simulate CreateJsonFieldMap
var map = new Dictionary<string, (string Name, string JsonName)>();
foreach (var field in fieldsInOrder)
{
    map[field.Name] = field;
    map[field.JsonName] = field;
}

// show
foreach (var (key, value) in map.OrderBy(x => x.Key))
{
    Console.WriteLine($"{key}: field {value.Name}");
}
```

IMO the correct implementation is more like:

```
var map = new Dictionary<string, (string Name, string JsonName)>();
foreach (var field in fieldsInOrder)
{
    map[field.Name] = field;
}
foreach (var field in fieldsInOrder)
{
    map[field.JsonName] = field;
}
```

Reply:
CreateJsonFieldMap replicates is what `Google.Protobuf` does. I'll update the comment here to be more accurate.

Reply:
@JamesNK that just kicks the can down the road :)

I guess the real question/test here is: what is the *intended* implementation, because I stand by the statement that this implementation looks demonstrably wrong; and I will grant that "intended" crosses over with "damn, we shipped it", but has this been compared/contrasted to e.g. the Java or C++ implementations?

But: this is probably better asked as an issue on protobuf, so: I'll go do that

Reply:
https://github.com/protocolbuffers/protobuf/issues/11987

Reply:
Thanks. I added info to the issue about what other implementations do.

It might take a while for Protobuf team to decide what to do. I'd rather not hold up this PR to wait for them. If Protobuf makes a change, then we can react to it then (both here and in the JSON deserializer)

Reply:
Yup; totally agree